File: Leetcode/Trees/balanced-binary-tree.java
// https://leetcode.com/problems/balanced-binary-tree/

// time - O(n)
// space - O(H) worst - O(n)
class Solution {
  boolean balanced = true;
  public boolean isBalanced(TreeNode root) {
    helper(root);
    return balanced;
  }
  
  public int helper(TreeNode root) {
    if(root == null) return 0;
    
    if(root.left == null && root.right == null) {
      return 1;
    }
    
    int leftHeight = helper(root.left);
    int rightHeight = helper(root.right);
      
    if(Math.abs(leftHeight - rightHeight) > 1) {
      balanced = false;
    }
      
    return 1 + Math.max(leftHeight, rightHeight);
  }
}

File: Blind75/Trees/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Time - O(N)
# Space - O(N)
class Solution:
    balanced = True

    def getHeight(self, root):
        if root == None:
            return 0
        
        leftSubtreeHeight = 1 + self.getHeight(root.left)
        rightSubtreeHeight = 1 + self.getHeight(root.right)

        if abs(leftSubtreeHeight - rightSubtreeHeight) > 1:
            self.balanced = False

        return max(leftSubtreeHeight, rightSubtreeHeight)
    
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self.getHeight(root)
        return self.balanced

# Without the global variable
class Solution:
    def getHeight(self, root):
        if root == None:
            return 0
        
        # propogate -1 up to the parent if we already know that the tree is not balanced somewhere down
        leftSubtreeHeight = self.getHeight(root.left)
        if leftSubtreeHeight == -1:
            return -1
        
        rightSubtreeHeight = self.getHeight(root.right)
        if rightSubtreeHeight == -1:
            return -1

        if abs(leftSubtreeHeight - rightSubtreeHeight) > 1:
            return -1

        return 1 + max(leftSubtreeHeight, rightSubtreeHeight)
    
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.getHeight(root) != -1
